### PR TITLE
[mono] Implement public hot reload API

### PR DESCRIPTION
**Pull request by lambdageek**
Also add ApplyUpdateSdb that takes byte[] arguments for use with
mono/debugger-libs as its awkward to call a ROS<byte> method from the debugger.

Contributes to https://github.com/dotnet/runtime/issues/45689

This is the mono counterpart of #48366

### CODE REVIEW

**Comment by lambdageek** [comment]
Probably going to conflict with the CoreCLR PR since we're both modifying the `ref/` assembly, but that's ok, I'll rebase after the other PR goes in.

**Comment by lambdageek** [comment]
@joj  once this is merged, (and there's a new nightly runtime installer) you will need to update your prototype to use
```csharp
System.Reflection.Metadata.AssemblyExtensions.ApplyUpdateSdb(Assembly assembly, byte[] metadataDelta, byte[] ilDelta, byte[] pdbDelta)
```

File: src/libraries/System.Runtime.Loader/ref/System.Runtime.Loader.cs
Comment (stephentoub) [inline src/libraries/System.Runtime.Loader/ref/System.Runtime.Loader.cs:13]: I assume this is just temporary as https://github.com/dotnet/runtime/pull/48366 is adding this as well (but slightly different).

File: src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (stephentoub) [inline src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:28]: Once https://github.com/dotnet/runtime/pull/48366 is in, can we refactor this to have most of this method in the shared partition?  Seems like the main difference is just the name of the ApplyUpdate_internal helper; we should be able to name the helpers the same and have just those pieces in the individual runtime partitions.

File: src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (marek-safar) [inline src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:55]: Why is this method needed?

File: src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (marek-safar) [inline src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:42]:
```suggestion
		    if (assembly is not RuntimeAssembly runtimeAssembly)
		    {
			    if (assembly is null) throw new ArgumentNullException(nameof(assembly));
			    throw new ArgumentException(SR.Argument_MustBeRuntimeAssembly);
		    }
```

File: src/libraries/System.Runtime.Loader/ref/System.Runtime.Loader.cs
Comment (lambdageek) [reply src/libraries/System.Runtime.Loader/ref/System.Runtime.Loader.cs:13]: Yep.  Once the other PR goes in, this will be removed.

File: src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (lambdageek) [reply src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:55]: I don't know how to call a method taking a `ReadOnlySpan<byte>` argument using debugger-libs.

Comment (marek-safar) [reply src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:50]: Shouldn't this code be in debugger-libs when ApplyUpdate is public API?

Comment (lambdageek) [reply src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:50]: The issue is that we don't want to add a debugger-libs command for hot reload (we could do it, but previously there didn't seem to be a need).  So we're using the existing "create a byte[] blob" and "invoke a function" debugger-libs commands.  The "invoke a function" debugger-libs command can't use `ROS<byte>` arguments - at least I don't know how to do it.

So there are three options:
1. Call a function with better arguments (what i'm doing in this PR)
2. or, add a soft debugger protocol command for applying a hot reload delta.
3. Add a way to pass `ref struct` arguments when invoking using the debugger.

Comment (marek-safar) [reply src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:50]: Why cannot we do "create a byte[] blob" and "invoke a function" inside debugger-libs? Either way, this method cannot be in a permanent descriptor or the hot reload will always ship with release config

Comment (lambdageek) [reply src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:50]: We will probably need to add a new command to the debugger protocol, then.

Can I keep this for now?

Comment (marek-safar) [reply src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:54]: I'd rather not as it's gonna screw our size tracking numbers

Comment (lambdageek) [reply src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:54]: Discussed offline: this method is (temporarily) replacing an existing LoadMetadataUpdate private API so the size tracking isn't affected by this PR.  I added an issue to switch to debugger libs and remove ApplyUpdateSdb.  But we're going to keep it for now to unblock downstream work

**Comment by lambdageek** [comment]
@stephentoub  I rebased on master picking up @mikem8361 's changes and moved the argument checking to the shared area.  Would you mind reviewing again?

@marek-safar leaving ApplyUpdateSdb as we discussed for now.

File: src/libraries/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:28]: I think it is contra-productive to be splitting a method into two to share 4 lines of trivial argument checking.

**Comment by mikem8361** [comment]
@lambdageek, are you going to enable the invalid arg testing in runtime\src\libraries\System.Runtime.Loader\tests\AssemblyExtensionsTest.cs? By removing the "SkipOnMono" and "ActiveIssue" attributes?

FYI, to build/run the test on Windows:
```
call c:\runtime\dotnet.cmd build /t:Test /p:XunitMethodName=System.Reflection.Metadata.AssemblyExtensionsTest.ApplyUpdateInvalidParameters C:\runtime\src\libraries\System.Runtime.Loader\tests\System.Runtime.Loader.Tests.csproj
```
On Linux/MacOS:
```
$HOME/runtime/dotnet.sh build /t:Test /p:XunitMethodName=System.Reflection.Metadata.AssemblyExtensionsTest.ApplyUpdateInvalidParameters $HOME/runtime/src/libraries/System.Runtime.Loader/tests/System.Runtime.Loader.Tests.csproj
```

**Comment by lambdageek** [comment]
Enabled the `ApplyUpdateInvalidParameters` test for Mono on all platforms.  ActiveIssue for CoreCLR on non-Windows. /cc @mikem8361

**Comment by lambdageek** [comment]
It's unfortunate that `PlatformSpecificAttribute` doesn't have a runtime parameter.  ActiveIssue isn't really a great fit

File: src/libraries/System.Runtime.Loader/tests/AssemblyExtensionsTest.cs
Comment (mikem8361) [inline src/libraries/System.Runtime.Loader/tests/AssemblyExtensionsTest.cs:15]: I found out the hard way but ActiveIssue doesn't actually affect whether the test runs.  The test is going to fail on Linux/MacOS on coreclr

**Review by mikem8361 — DISMISSED** [review]

File: src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (lambdageek) [reply src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs]: Getting some conflicting guidance here https://github.com/dotnet/runtime/pull/48380#discussion_r578846729 :smile: 

But actually maybe your suggestion was to just leave the QCall / InternallCall runtime-specific and to share the invocations.  Unfortunately that won't work today - the arguments are a bit different for each runtime.

File: src/libraries/System.Runtime.Loader/tests/AssemblyExtensionsTest.cs
Comment (lambdageek) [reply src/libraries/System.Runtime.Loader/tests/AssemblyExtensionsTest.cs:15]: I'm not totally convinced.  Just tried
```
        [Fact]
        [ActiveIssue("https://github.com/dotnet/runtime/issues/45689", platforms: ~TestPlatforms.Windows, runtimes: TestRuntimes.Mono)]
        public static void FailingTest()
        {
                Assert.True(false);
        }
```

and it skipped the test on Linux using Mono.  and when I changed to `TestPlatforms.CoreCLR` running on Mono on Linux ran the test and failed.

Comment (stephentoub) [reply src/libraries/System.Runtime.Loader/tests/AssemblyExtensionsTest.cs:15]: > I found out the hard way but ActiveIssue doesn't actually affect whether the test runs. 

It should.  If it doesn't, that's a bug somewhere.

File: src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (stephentoub) [reply src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs]: > maybe your suggestion was to just leave the QCall / InternallCall runtime-specific and to share the invocations

Yeah, that's what I was hoping for.  If it's not possible, ok.

File: src/libraries/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (lambdageek) [reply src/libraries/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:28]: Yep, will revert.  I misunderstood what Stephen was suggesting.

File: src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.Mono.cs
Comment (stephentoub) [inline src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.Mono.cs:20]: Should these strings be in the resx?

Comment (lambdageek) [reply src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.Mono.cs]: fixed

File: src/libraries/System.Runtime.Loader/tests/AssemblyExtensionsTest.cs
Comment (lambdageek) [reply src/libraries/System.Runtime.Loader/tests/AssemblyExtensionsTest.cs:15]: runtime (Libraries Test Run release coreclr OSX x64 Debug) passed, so looks like it's working.

Comment (mikem8361) [reply src/libraries/System.Runtime.Loader/tests/AssemblyExtensionsTest.cs:15]: I must have had things backwards or something.

**Comment by lambdageek** [comment]
Is there anything else that should be done before this is merged?  If not could someone approve.

File: src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (stephentoub) [inline src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:41]: Not sure if this message was left in the code on purpose or should be extracted to the .resx.

File: src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (stephentoub) [inline src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:28]: The `= default` should be removed I believe.

File: src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs
Comment (stephentoub) [inline src/mono/System.Private.CoreLib/src/System/Reflection/Metadata/AssemblyExtensions.cs:54]: Is the `= null` needed here?

**Review by stephentoub — APPROVED** [review]

**Pull request merged**
